events.v2/query/query.go

```diff
@@ -0,0 +1,215 @@
+}
+
+func (q *Query) Matches(tags map[string]interface{}) bool {
+	for tag, operatorAndValue := range q.parser.queryMap.Map() {
+		value, ok := tags[tag]
+		if !ok { // no tag in tags matching a tag from query
+			return false
+		}
+		switch operatorAndValue.Value.Kind() {
+		case reflect.Struct: // time
+			v, ok := value.(time.Time)
+			if !ok {
+				return false
+			}
+			switch operatorAndValue.Op {
+			case opLessEqual:
+				if v.Before(operatorAndValue.Value.Interface().(time.Time)) || v.Equal(operatorAndValue.Value.Interface().(time.Time)) {
+					return true
+				}
+			case opGreaterEqual:
+				if v.Equal(operatorAndValue.Value.Interface().(time.Time)) || v.After(operatorAndValue.Value.Interface().(time.Time)) {
+					return true
+				}
+			case opLess:
+				if v.Before(operatorAndValue.Value.Interface().(time.Time)) {
+					return true
+				}
+			case opGreater:
+				if v.After(operatorAndValue.Value.Interface().(time.Time)) {
+					return true
+				}
+			case opEqual:
+				if v.Equal(operatorAndValue.Value.Interface().(time.Time)) {
+					return true
+				}
+			}
+		case reflect.Float64:
+			v, ok := value.(float64)
+			if !ok {
+				return false
+			}
+			switch operatorAndValue.Op {
+			case opLessEqual:
+				if v <= operatorAndValue.Value.Interface().(float64) {
+					return true
+				}
+			case opGreaterEqual:
+				if v >= operatorAndValue.Value.Interface().(float64) {
+					return true
+				}
+			case opLess:
+				if v < operatorAndValue.Value.Interface().(float64) {
+					return true
+				}
+			case opGreater:
+				if v > operatorAndValue.Value.Interface().(float64) {
+					return true
+				}
+			case opEqual:
+				if v == operatorAndValue.Value.Interface().(float64) {
+					return true
+				}
+			}
+		case reflect.Int:
+			v, ok := value.(int)
+			if !ok {
+				return false
+			}
+			switch operatorAndValue.Op {
+			case opLessEqual:
+				if v <= operatorAndValue.Value.Interface().(int) {
+					return true
+				}
+			case opGreaterEqual:
+				if v >= operatorAndValue.Value.Interface().(int) {
+					return true
+				}
+			case opLess:
+				if v < operatorAndValue.Value.Interface().(int) {
+					return true
+				}
+			case opGreater:
+				if v > operatorAndValue.Value.Interface().(int) {
+					return true
+				}
+			case opEqual:
+				if v == operatorAndValue.Value.Interface().(int) {
+					return true
+				}
+			}
+		case reflect.String:
+			v, ok := value.(string)
+			if !ok {
+				return false
+			}
+			switch operatorAndValue.Op {
+			case opEqual:
+				if v == operatorAndValue.Value.String() {
+					return true
+				}
+			case opContains:
+				if strings.Contains(v, operatorAndValue.Value.String()) {
+					return true
+				}
+			}
+		}
+	}
+
+	return false
+}
+
+type operator uint8
+
+const (
+	opLessEqual operator = iota
+	opGreaterEqual
+	opLess
+	opGreater
+	opEqual
+	opContains
+)
+
+type operatorAndValue struct {
+	Op        operator
+	ValueType reflect.Type
+	Value     reflect.Value
+}
+
+type queryMap struct {
```

Comment:
remove the map, just go over the AST from `query.peg.go` (run a benchmark to see if is not slower)